File: apps/frontend/services/admin.service.ts
import { API_BASE_URL, type ApiResponse } from "@/lib/auth";

export interface ActiveHackathonSummary {
  id: string;
  title: string;
  start_time: string | null;
  end_time: string | null;
  problem_count: number;
}

export interface SubmissionStatusBreakdown {
  pending: number;
  running: number;
  completed: number;
  failed: number;
}

export interface TopScoreEntry {
  user_id: string;
  full_name: string;
  total_points: number;
}

export interface AdminDashboardData {
  total_students: number;
  total_problems: number;
  total_submissions: number;
  total_violations: number;
  active_hackathon: ActiveHackathonSummary | null;
  submission_status_breakdown: SubmissionStatusBreakdown;
  top_scores: TopScoreEntry[];
}

async function request<T>(path: string, init?: RequestInit): Promise<ApiResponse<T>> {
  const response = await fetch(`${API_BASE_URL}${path}`, {
    ...init,
    credentials: "include",
    headers: {
      "Content-Type": "application/json",
      ...(init?.headers ?? {}),
    },
  });

  const body = (await response.json()) as ApiResponse<T>;
  if (!response.ok) {
    throw new Error(body.message || "Request failed");
  }
  return body;
}

export function getAdminDashboard() {
  return request<AdminDashboardData>("/admin/dashboard", { method: "GET" });
}

File: apps/backend/app/services/admin_service.py
from sqlalchemy.orm import Session

from app.repositories.admin_repository import AdminRepository
from app.schemas.admin import (
    ActiveHackathonSummary,
    AdminDashboardResponse,
    SubmissionStatusBreakdown,
    TopScoreEntry,
)


class AdminService:
    def __init__(self, db: Session) -> None:
        self.db = db
        self.repo = AdminRepository(db)

    def get_dashboard(self) -> AdminDashboardResponse:
        active_hackathon = self.repo.get_active_hackathon()
        active_hackathon_summary = None
        if active_hackathon is not None:
            active_hackathon_summary = ActiveHackathonSummary(
                id=active_hackathon.id,
                title=active_hackathon.title,
                start_time=active_hackathon.start_time,
                end_time=active_hackathon.end_time,
                problem_count=self.repo.count_problems_for_hackathon(active_hackathon.id),
            )

        status_counts = self.repo.submission_status_counts()
        breakdown = SubmissionStatusBreakdown(
            pending=status_counts.get("PENDING", 0),
            running=status_counts.get("RUNNING", 0),
            completed=status_counts.get("COMPLETED", 0),
            failed=status_counts.get("FAILED", 0),
        )

        top_scores = [
            TopScoreEntry(user_id=user.id, full_name=user.full_name, total_points=points)
            for user, points in self.repo.top_scores()
        ]

        return AdminDashboardResponse(
            total_students=self.repo.count_students(),
            total_problems=self.repo.count_problems(),
            total_submissions=self.repo.count_submissions(),
            total_violations=self.repo.count_violations(),
            active_hackathon=active_hackathon_summary,
            submission_status_breakdown=breakdown,
            top_scores=top_scores,
        )
